injective_functions/staking/__init__.py

```diff
@@ -19,3 +20,89 @@ async def stake_tokens(self, validator_address: str, amount: str) -> Dict:
             amount=float(amount),
         )
         return await self.chain_client.build_and_broadcast_tx(msg)
+
+    async def compound_rewards(self, validator_address: str) -> Dict:
+        """
+        Compounds staking rewards by withdrawing them and restaking.
+        :param validator_address: The validator's address
+        :return: Transaction result
+        """
+        try:
+            if not validator_address.startswith("injvaloper"):
+                raise ValueError("Invalid validator address format")
+
+            # Step 1: Fetch the initial INJ balance
+            balance_response = await self.chain_client.client.get_bank_balance(
+                address=self.chain_client.address.to_acc_bech32(),
+                denom="inj"
+            )
+            initial_balance = Decimal(balance_response.balance.amount)
+
+            # Step 2: Withdraw rewards
+            withdraw_msg = self.chain_client.composer.msg_withdraw_delegator_reward(
+                delegator_address=self.chain_client.address.to_acc_bech32(),
+                validator_address=validator_address,
+            )
+            withdraw_response = await self.chain_client.build_and_broadcast_tx(withdraw_msg)
+
+            # Step 3: Wait for the balance to update
+            updated_balance = await self.wait_for_balance_update(old_balance=initial_balance, denom="inj")
+
+            # Step 4: Calculate the withdrawn rewards
+            rewards_to_stake = updated_balance - initial_balance
+            if rewards_to_stake <= 0:
+                if rewards_to_stake < 0:
+                    # Specific error for negative rewards
+                    return {
+                        "success": False,
+                        "error": f"Rewards ({rewards_to_stake}) are lower than gas fees, resulting in a negative net "
+                                 f"reward."
+                    }
+                # Generic error for zero rewards
+                return {"success": False, "error": "No rewards available to compound."}
+
+            # Step 5: Restake the rewards
+            delegate_msg = self.chain_client.composer.MsgDelegate(
+                delegator_address=self.chain_client.address.to_acc_bech32(),
+                validator_address=validator_address,
+                amount=rewards_to_stake / Decimal('1e18'),
```

Comment:
I haven't checked other parts of the code here but suggest creating a constant for the 1e18 decimal conversion

Reply:
I’ve defined the constant within the InjectiveStaking class for now, as it’s currently only being used in this context. If the need arises to use it more broadly in the future, we can consider moving it to a shared location for better accessibility.